components/client/src/test/unit/java/com/hotels/styx/client/retry/RetryNTimesTest.java

```diff
@@ -0,0 +1,117 @@
+        RetryNTimes retryNTimes = new RetryNTimes(1);
+
+        RetryPolicy.Context context = mock(RetryPolicy.Context.class);
+        when(context.currentRetryCount()).thenReturn(0);
+        when(context.lastException()).thenReturn(of(new TestException()));
+        RetryPolicy.Outcome retryOutcome = retryNTimes.evaluate(context,
+                mock(LoadBalancingStrategy.class), mock(LoadBalancingStrategy.Context.class));
+
+        assertThat(retryOutcome.shouldRetry(), equalTo(true));
+    }
+
+    @Test
+    public void shouldNotRetryBasedOnMaxAttemptsReached() {
+        RetryNTimes retryNTimes = new RetryNTimes(1);
+
+        RetryPolicy.Context context = mock(RetryPolicy.Context.class);
+        when(context.currentRetryCount()).thenReturn(1);
+        when(context.lastException()).thenReturn(of(new TestException()));
+        RetryPolicy.Outcome retryOutcome = retryNTimes.evaluate(context,
+                mock(LoadBalancingStrategy.class), mock(LoadBalancingStrategy.Context.class));
+
+        assertThat(retryOutcome.shouldRetry(), equalTo(false));
+    }
+
+    @Test
+    public void shouldNotRetryBasedOnWrongException() {
+        RetryNTimes retryNTimes = new RetryNTimes(1);
+
+        RetryPolicy.Context context = mock(RetryPolicy.Context.class);
+        when(context.currentRetryCount()).thenReturn(0);
+        when(context.lastException()).thenReturn(of(new RuntimeException()));
+        RetryPolicy.Outcome retryOutcome = retryNTimes.evaluate(context,
+                mock(LoadBalancingStrategy.class), mock(LoadBalancingStrategy.Context.class));
+
+        assertThat(retryOutcome.shouldRetry(), equalTo(false));
+    }
+
+    @Test
+    public void shouldReturnUnfilteredOrigin() {
+        RetryNTimes retryNTimes = new RetryNTimes(1);
+
+        RetryPolicy.Context context = mock(RetryPolicy.Context.class);
+        ConnectionPool connectionPool = mock(ConnectionPool.class);
+        LoadBalancingStrategy.Context lbContext = mock(LoadBalancingStrategy.Context.class);
+        LoadBalancingStrategy loadBalancingStrategy = mock(LoadBalancingStrategy.class);
+
+        when(context.currentRetryCount()).thenReturn(0);
+        when(context.lastException()).thenReturn(empty());
+        when(context.previousOrigins()).thenReturn(Collections.emptyList());
+
+        when(loadBalancingStrategy.vote(lbContext)).thenReturn(singleton(connectionPool));
+        RetryPolicy.Outcome retryOutcome = retryNTimes.evaluate(context,
+                loadBalancingStrategy, lbContext);
+
+        assertThat(retryOutcome.nextOrigin().get(), equalTo(connectionPool));
+    }
+
+    @Test
+    public void shouldReturnEmptyOriginList() {
+        RetryNTimes retryNTimes = new RetryNTimes(1);
+
+        RetryPolicy.Context context = mock(RetryPolicy.Context.class);
+        ConnectionPool connectionPool = mock(ConnectionPool.class);
+        LoadBalancingStrategy.Context lbContext = mock(LoadBalancingStrategy.Context.class);
```

Comment:
The mock setup adds lots of verbosity in the tests in this class. Please consider:
1) adding a helper method to set up context
2) convert the mocks like `connectionPool`, `lbContext`, `loadBalancingStrategy` into fields and initialise in test setup method before each test.